topology/nhlt/intel/dmic-nhlt.c

```diff
@@ -0,0 +1,1002 @@
+				return -EINVAL;
+			dai_index_t = int_val;
+		}
+
+		if (!strcmp(id, "num_pdm_active")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			num_pdm_active = int_val;
+		}
+
+		if (!strcmp(id, "fifo_word_length")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			fifo_word_length = int_val;
+		}
+
+		if (!strcmp(id, "clk_min")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			clk_min = int_val;
+		}
+
+		if (!strcmp(id, "clk_max")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			clk_max = int_val;
+		}
+
+		if (!strcmp(id, "duty_min")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			duty_min = int_val;
+		}
+
+		if (!strcmp(id, "duty_max")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			duty_max = int_val;
+		}
+
+		if (!strcmp(id, "sample_rate")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			sample_rate = int_val;
+		}
+
+		if (!strcmp(id, "unmute_ramp_time_ms")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			unmute_ramp_time_ms = int_val;
+		}
+	}
+
+	return dmic_set_params(tplg_pp, dai_index_t, driver_version, io_clk, num_pdm_active,
+			       fifo_word_length, clk_min, clk_max, duty_min, duty_max, sample_rate,
+			       unmute_ramp_time_ms);
+}
+
+static int set_pdm_data(struct tplg_pre_processor *tplg_pp,
+			snd_config_t *cfg)
+{
+	snd_config_iterator_t i, next;
+	snd_config_t *class_cfg;
+	int mic_a_enable;
+	int mic_b_enable;
+	snd_config_t *n;
+	const char *id;
+	int polarity_a;
+	int polarity_b;
+	int clk_edge;
+	long int_val;
+	int ctrl_id;
+	int skew;
+	int ret;
+
+	/* get default values */
+	ret = snd_config_search(tplg_pp->input_cfg, "Class.Base.pdm_config", &class_cfg);
+	if (ret < 0)
+		return -EINVAL;
+
+	/* set default values */
+	snd_config_for_each(i, next, class_cfg) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)
+			continue;
+
+		if (!strcmp(id, "ctrl_id")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			ctrl_id = int_val;
+		}
+
+		if (!strcmp(id, "mic_a_enable")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			mic_a_enable = int_val;
+		}
+
+		if (!strcmp(id, "mic_b_enable")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			mic_b_enable = int_val;
+		}
+
+		if (!strcmp(id, "polarity_a")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			polarity_a = int_val;
+		}
+
+		if (!strcmp(id, "polarity_b")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			polarity_b = int_val;
+		}
+
+		if (!strcmp(id, "clk_edge")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			clk_edge = int_val;
+		}
+
+		if (!strcmp(id, "skew")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			skew = int_val;
+		}
+	}
+
+	/* set object values */
```

Comment:
why do you need to set class values? The object should have inherited class values already no?

Reply:
so are the class values already set at this point in auto_attr processing?

Reply:
I would think so